soccer/gameplay/skills/capture.py

```diff
@@ -1,270 +1,241 @@
+                             name='collector',
+                             required=True,
+                             priority=10)
+
+    def on_enter_completed(self):
+        self.probably_held_cnt = Capture.PROBABLY_HELD_START
+        self.remove_all_subbehaviors()
+
+    def execute_completed(self):
+        self.update_held_cnt()
+
+        self.robot.disable_avoid_ball()
+        self.robot.set_dribble_speed(constants.Robot.Dribbler.StandardPower)
+
+    def update_held_cnt(self):
+        if (self.robot is None):
+            return
+
+        # If we hold the ball, increment up to max
+        # if not, decrement to 0
+        if (evaluation.ball.robot_has_ball(self.robot)): #self.robot.has_ball()):
+            self.probably_held_cnt = min(self.probably_held_cnt + 1, Capture.PROBABLY_HELD_HISTORY_LENGTH)
         else:
-            self.robot.set_avoid_ball_radius(Capture.CoarseApproachAvoidBall)
+            self.probably_held_cnt = max(self.probably_held_cnt - 1, 0)
 
-        self.robot.move_to(pos)
-        main.system_state().draw_circle(self.lastApproachTarget,
-                                        constants.Ball.Radius,
-                                        constants.Colors.White, "Capture")
+    @staticmethod
+    def role_assignment_cost(robot):
+        # Estimate the time it takes to intercept the ball
 
-    def on_exit_coarse_approach(self):
-        self.lastApproachTarget is None
+        # Use straight line distance to closest point along line
+        # Then movement down the line
 
-    def execute_fine_approach(self):
-        self.robot.disable_avoid_ball()
-        self.robot.set_dribble_speed(self.dribbler_power)
+        # Represents ball movement line
+        ball_move_line = robocup.Line(main.ball().pos, main.ball().pos + main.ball().vel * 10)
 
-        bot2ball_dir = (main.ball().pos - self.robot.pos).normalized()
+        # Closest point to robot on ball movement line
+        ball_line_intercept = ball_move_line.nearest_point(robot.pos)
 
-        approach = self.bot_to_ball() * Capture.FineApproachDistanceMultiplier + \
-                    bot2ball_dir * Capture.FineApproachMinDeltaSpeed + \
-                    main.ball().vel * Capture.FineApproachBallSpeedMultiplier
-        if (approach.mag() > 1):
-            approach = approach.normalized() * 1
-        self.robot.set_world_vel(approach)
+        ball_to_robot = robot.pos - main.ball().pos
+        ball_to_intercept = ball_line_intercept - main.ball().pos
+        robot_to_intercept = ball_line_intercept - robot.pos
 
-    def on_enter_delay(self):
-        self.start_time = time.time()
+        # If we are actually in front of the ball angle wise
+        # OR close enough that it could be a bounce off the robot due to noisy vision
+        in_front = ball_move_line.delta().angle_between(ball_to_robot) < math.pi/2 or \
+                   ball_to_robot.mag() < 1.5*(constants.Robot.Radius + constants.Ball.Radius)
```

Comment:
Are you ok with the fact that the ball could be moving away from the robot here?